File: app/src/main/java/com/keiferstone/dndinitiativetracker/CharacterStorage.java
package com.keiferstone.dndinitiativetracker;

import android.content.Context;
import android.content.SharedPreferences;
import android.support.annotation.NonNull;

import com.google.gson.Gson;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

class CharacterStorage {
    private static final String CHARACTER_STORAGE = "characterStorage";

    private SharedPreferences sharedPreferences;

    CharacterStorage(Context context) {
        sharedPreferences = context.getSharedPreferences(CHARACTER_STORAGE, Context.MODE_PRIVATE);
    }

    @NonNull
    List<Character> loadAllCharacters() {
        List<Character> characters = new ArrayList<>();
        Map<String, ?> characterPrefs = sharedPreferences.getAll();
        for (Object characterJson : characterPrefs.values()) {
            if (characterJson instanceof String) {
                characters.add(new Gson().fromJson((String) characterJson, Character.class));
            }
        }
        return characters;
    }

    void saveCharacter(@NonNull Character character) {
        sharedPreferences.edit().putString(character.getId(), new Gson().toJson(character)).apply();
    }

    void deleteCharacter(@NonNull Character character) {
        sharedPreferences.edit().remove(character.getId()).apply();
    }
}

File: app/src/main/java/com/keiferstone/dndinitiativetracker/CharacterStorage.kt
package com.keiferstone.dndinitiativetracker

import android.content.Context
import android.content.SharedPreferences
import com.google.gson.Gson

internal class CharacterStorage(context: Context) {
    private val sharedPreferences: SharedPreferences
            = context.getSharedPreferences("characterStorage", Context.MODE_PRIVATE)

    fun loadAllCharacters(): List<Character> {
        return sharedPreferences.all.values
                .filterIsInstance<String>()
                .map { Gson().fromJson(it, Character::class.java) }
    }

    fun saveCharacter(character: Character) {
        sharedPreferences.edit().putString(character.id, Gson().toJson(character)).apply()
    }

    fun deleteCharacter(character: Character) {
        sharedPreferences.edit().remove(character.id).apply()
    }
}